Add support for Todoist platform

Jay2645 opened the pull request:
## Description:
Support for Todoist task management (https://todoist.com).

Todoist breaks task management down into 2 structures:
* Tasks, which are "things" you can do and can mark as being done
* Projects, which are essentially containers for tasks.

In Home Assistant, each project gets its own calendar. The start date of the calendar is set to `now`, while the end date is set to the due date -- the logic for this being that most tasks you set on here are tasks you could theoretically do *now*. There are some exceptions (like taking someone to the airport), but a good percentage of the things on your to-do list could be done right now if you wanted to. Because of that, the calendar takes a simple state -- `on` if you have a task in it that's not marked as completed, or `off` if you have no tasks at all in that project.

The calendar keeps track of what your most "important" task in that project is, governed by a number of factors specified in the header of the `.py` file. You can access all the tasks in a project via a template if you wanted to, and they'd all be sorted from most important to least important.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3283

## Example entry for `configuration.yaml` (if applicable):
```yaml
calendar:
  - platform: todoist
    token: !secret todoist_token
    custom_projects:
      - name: 'All Projects'
      - name: 'Due Today'
        due_date_days: 0
      - name: 'Due This Week'
        due_date_days: 7
      - name: 'Math Homework'
        labels:
          - Homework
        include_projects:
          - Mathematical Structures II
          - Calculus II
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

~~If the code does not interact with devices:~~
  - ~~Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**~~
  - ~~Tests have been added to verify that the new code works.~~

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Jay2645 committed:
Added basic Todoist support

Creating a new platform for Todoist - https://todoist.com

Jay2645 committed:
Added more robust support for creating new custom projects.

This means you can now specify things such as 'all tasks due today', 'all tasks due this week', etc.

Jay2645 committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into add-todoist

Jay2645 committed:
Changed logging from warning to info.

Jay2645 committed:
Added label and comment support.

Jay2645 committed:
Added support for overdue tasks.

Jay2645 committed:
Changed logging to info instead of warning; fixed labels.

Jay2645 committed:
Added ability to filter projects by name.

Jay2645 committed:
Rename 'extra_projects' to 'custom_projects'.

Jay2645 committed:
Updated code to follow proper HASS style guidelines.

Jay2645 committed:
Got new_task service running.

Jay2645 committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into add-todoist

Jay2645 committed:
Update .coveragerc.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,604 @@
+        return attributes
+
+
+class TodoistProjectData(object):
```
This should probably be broken out into a python package and published on pypi. Is there no pypi package already for todoist?

Jay2645 replied:
[There is](https://pypi.python.org/pypi/todoist-python), but I didn't want to add any additional dependencies to the project, since Todoist has a REST API which accomplishes the same exact tasks. I considered adding their API as a dependency when creating the "Add Task" service, but in the end I felt that it wasn't worth it. When I was designing this, I had a misconception that using the API *required* you [to use their OAuth implementation](https://developer.todoist.com/sync/v7/#oauth), which would make my life much more difficult in terms of authenticating the user. That was another thing which pushed me towards using the REST API, although now that I'm looking back at it I see there's actually a way to authenticate with tokens after all.

That being said, half the code highlighted here is just me overriding HA's default Calendar implementation. If I were to move to the PyPI package, most of this code would largely be the same (since it's modifying the Todoist "device" based on the `TodoistProjectData` and not actually doing any real logic of its own). I would probably be able to get rid of the `TodoistProjectData` object by moving to PyPI, but I don't want to introduce more dependencies than I have to, especially when there's an "easy" way around it. I would also still need to "score" each task in order to display them properly in Home Assistant, something which is currently done in `TodoistProjectData`. If you'd like, though, I'm okay with doing that.

Jay2645 replied:
[There is](https://pypi.python.org/pypi/todoist-python), but I didn't want to introduce any dependencies. When I first started designing the platform, I had the misconception after looking at [their Python API](https://developer.todoist.com/sync/v7/) that I would need to do everything via OAuth, which is obviously more of a challenge than just having the user specify a token in their config. Looking back at it now, I see that there is actually a way to do things with a token, but I'm still adverse to introducing extra dependencies.

That being said, I would still need *some* Todoist logic somewhere. This class has the scoring algorithm, which could be moved to the actual device. But (more importantly), it also supports "custom" projects -- i.e., where the user can ask to grab everything in the next week, or all things tagged "Homework", or what have you. That stuff comes from the config file, and it'll all have to be kept track of locally (otherwise users may wind up accidentally creating a new project as soon as the API syncs, from what I've gathered in a cursory examination).

This class also "converts" the data from Todoist to something that Home Assistant can process easier -- as an example, the due date logic around line 400 or so. Sure, I'd be able to get an actual datetime object from the API instead of relying upon `strptime`, but I would still need to "translate" that to the end time for a Calendar in Home Assistant. This also ensures that every single task has a start and end date, that there's always a list of comments (even if it's empty), etc. By default, if something is empty, there's just no entry for it in the API. The idea is to do as much logic as makes sense here, in order to cut down on the amount of logic contained within the actual device. I can probably move the logic for determining if something is due today or overdue into the device, but at the time it seemed fitting just to do it here, since I was already working with the data.

The end result is I would need some class which looks vaguely like this, except I would have references to Todoist's PyPI API calls instead of using `requests`. I'd still need to "score" each task, since Todoist sorts by date added, not by actual priority. I'd still need to "translate" a lot of the Todoist-specific things to what the Calendar device expects, but perhaps this can be done on the device end. Most importantly, though, I'd still need to keep a list of "local" projects for those user-defined devices that can be specified in the config. That'd be where most of the logic would come from.

I can do it, if you'd like me to. Doing so would give me access to Todoist Reminders, so I could probably expose the Reminder functionality built-in to Todoist to Home Assistant (right now, I can't access Reminders via the REST API -- just Tasks and their due dates). It would be a fairly easy change overall, but I'm not sure whether it would cut down on the amount of platform-specific logic unless I got rid of the custom user-defined projects (which is where a lot of the logic comes from currently).

MartinHjelmare replied:
Introducing a new dependency is not a problem, but rather part of the ground rules for home assistant:
https://home-assistant.io/developers/code_review_platform/#6-communication-with-devicesservices
https://home-assistant.io/developers/code_review_platform/#1-requirements

Having an interface class or callback function between the pypi package API and the home assistant entity API is ok, but we try to minimize it as much as possible. Usually this class/function is responsible for fetching/accepting new data into home assistant.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,604 @@
+CONF_PROJECT_WHITELIST = 'include_projects'
+CONF_PROJECT_LABEL_WHITELIST = 'labels'
+
+CONFIG_SCHEMA = vol.Schema({
```
Platforms should implement `PLATFORM_SCHEMA`.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,604 @@
+                task_data['due_datetime'] = call.data['due_datetime']
+            elif 'due_date' in call.data:
+                task_data['due_date'] = call.data['due_date']
+        except KeyError:
```
The service schema should validate `call.data` to make this error impossible.

MartinHjelmare reviewed (changes requested):
I think you should use a pypi package for the todoist specific logic.

Jay2645 committed:
Remove old try-catch block.

This is left over from before we validated the inputs using the service schema.

Jay2645 committed:
Updated to use PLATFORM_SCHEMA.

Jay2645 committed:
Updated component to use Todoist API.

Jay2645 committed:
Removed commented-out code.

This also removes functionality regarding finding out how many comments a task has.
This functionality may be added back in the future.

Jay2645 committed:
Clarified TodoistProjectData, removed fetching comments.

Jay2645 committed:
Fixed bug where projects were grabbing all tasks.

Jay2645 committed:
Fixed bug where due dates were being ignored.

Jay2645 committed:
Removed debug logging.

Jay2645 commented:
@MartinHjelmare I've made the changes you requested.

* I changed from implementing `CONFIG_SCHEMA` to implementing `PLATFORM_SCHEMA`; I'm not 100% sure I'm doing things right, but it seems to work in Home Assistant. 

* `call.data` was already being validated; the `try-catch` block you pointed out was from an older version before I implemented validation. I went ahead and just removed it.

* In the end, I wasn't able to get rid of the `TodoistProjectData` interface class; I forgot that the Calendar platform itself requires an interface class [with an `update` function](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/calendar/__init__.py#L137) and [an `event` object](https://github.com/home-assistant/home-assistant/blob/3b2bf1d567d8debc80284a0c417c476c3acba6af/homeassistant/components/calendar/__init__.py#L153), both of which get defined in the `TodoistProjectData` interface class. This pattern is seen in [the Demo Calendar component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/calendar/demo.py#L29) as well as [the Google Calendar component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/calendar/google.py#L52).

Let me know if you need any more changes!

Jay2645 committed:
Fixed linter errors.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,595 @@
+            has the same priority as our current event, but is due earlier
+            in the day, select it
+        """
+        if len(project_tasks) > 0:
```
You already make this check before you call this method. I suggest you assume you get a list and just iterate over that.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,595 @@
+            # Start at the end of the list, so if tasks don't have a due date
+            # the newest ones are the most important.
+
+            event = project_tasks[len(project_tasks) - 1]
```
```py
event = project_tasks[-1]
```

MartinHjelmare reviewed (changes requested):
My comments are mostly about three themes:

- Use the schema validation as much as possible to get good default values and simplify later code.
- Optimize checks and iterations using (list) comprehension.
- Only log at debug level.

Jay2645 committed:
Fixed Todoist docstring to be in line with HASS' style rules.

Jay2645 committed:
Organized imports.

Jay2645 committed:
Fixed voluptuous schema.

Jay2645 committed:
Moved ID lookups into .

Jay2645 committed:
Moved ID lookups into setup_platform.

Jay2645 committed:
Cleaned up setup_platform a bit.

Jay2645 committed:
Cleaned up Todoist service calls.

Jay2645 committed:
Changed debug logging level.

Jay2645 committed:
Fixed issue with configuration not validating.

Jay2645 committed:
Changed from storing the token to storing an API instance.

Jay2645 commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,595 @@
+        self._token = token
+        self._name = project_data['name']
+        # If no ID is defined, fetch all tasks.
+        if project_data['id'] is not None:
```
So as it turns out, that's not a dict: it's a `Project` object from the Todoist API. I had been operating under the assumption that it was a dict, [since that's what their API seems to imply.](https://developer.todoist.com/sync/v7/#projects) I'm not sure where exactly that `Project` object is stored in their API; `from todoist import Project` doesn't seem to work, and they don't really have it documented much in their API. I can create a new project using their API calls, but then it'd be synchronized with their server.

All I *really* need from their `Project` object is the name and ID of the project, so I can either handle it with a `try-except` block here to catch the missing key or just have the method accept a flat-out dict instead of a `Project` object. I'll likely just go with the latter, since it seems less hacky.

Jay2645 committed:
Use dict instead of Project object.

Jay2645 committed:
Updated to use list comprehension where possible.

Jay2645 committed:
Fixed linter errors.

Jay2645 commented:
Alright, I believe all of your changes should be good to go! Let me know if you have any more!

MartinHjelmare reviewed (changes requested):
Good work! Some comments below. An optional improvement would be to define string constants at the top of the module for the magic strings, eg:
```py
PROJECTS = 'projects'
LABELS = 'labels'
```
and use those where needed in the code. It will make typo risk less, look cleaner, and be easier to change a string in the future.

Jay2645 commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -0,0 +1,488 @@
+            if label['id'] in data['labels']]
+
+        whitelist = self._label_whitelist
+        if any(label in task['labels'] for label in whitelist):
```
I had tried that (you had suggested that change last time), and for whatever reason everything was inverted. I'm not exactly sure why it works, but it works.

MartinHjelmare replied:
Looking at this again. Shouldn't all labels in `task[LABELS]` be on the whitelist to be ok? Otherwise we should return `None`. Then I think it should be:
```py
if not all(label in self._label_whitelist) for label in task[LABELS]:
    return None
```

Jay2645 replied:
@MartinHjelmare 
So, here's the results with `if not all`:

![if not all](https://user-images.githubusercontent.com/2058763/30345070-a356d92a-97b8-11e7-87fa-cdc20a0f6648.png)

And here is `if any`:

![if any](https://user-images.githubusercontent.com/2058763/30345105-c3dc330c-97b8-11e7-8e8b-7cf8b6d534aa.png)

`if any` gives the correct results (with the label whitelist working as an `or` statement, [as documented here](https://github.com/home-assistant/home-assistant.github.io/pull/3283/files#diff-862da73f775be0e9540cb46d2513aa26R40)).

`if not all(label for self._label_whitelist) for label in task[LABELS]` causes *any* task with a label to be rejected for some reason. You can see that in the first screenshot -- the project "Mathematical Structures II only consists of tasks with the label "Homework" or "School" (or both), so `calendar.mathematical_structures_ii` rejects all of them. This change propagates down to my "custom" calendars specified in my configuration: `calendar.math_homework` and `calendar.due_today` are only taking into account tasks which don't have labels. You can see `calendar.movies_to_watch` is not affected, as none of the tasks there have labels.

The end result is only tasks with exactly 0 labels are accepted, with the whitelist not being taken into account at all.

`if any(label in task[LABELS] for label in self._label_whitelist)` returns the correct values, but I'm not sure why. I agree with you that there should be a `not` in there somewhere, but a `not` gives me inverted inputs.

I'll take a look at what's going on exactly and see if I can fix it. There must be some logic error somewhere. Maybe `if any(label in task[LABELS] for label in self._label_whitelist)` is just never returning true, and thus never tossing out anything?

Jay2645 committed:
Use constants instead of literals.

Jay2645 committed:
Changed logging to use old-style string formatting.

Jay2645 committed:
Removed unneeded caching.

Jay2645 commented:
@MartinHjelmare Your changes should be live now.

Jay2645 committed:
Added comments explaining 'magic' strings.

MartinHjelmare commented:
@Jay2645 see my comment here https://github.com/home-assistant/home-assistant/pull/9236#discussion_r138329606. Let me know if I'm misunderstanding the whitelist.

Jay2645 committed:
Fixed bug where labels were always on the whitelist.

Jay2645 commented:
Fixed it. The whitelist is an *or*, not an *and*, (see [here](https://github.com/home-assistant/home-assistant.github.io/pull/3283/files#diff-862da73f775be0e9540cb46d2513aa26R40)), but that wasn't the issue after all.

The issue is I was comparing a list of user-defined label names to a list of the names of labels Todoist reports as being part of the task. The names Todoist reported were all lowercase; the names the user defined could be uppercase or lowercase. It was comparing `'Homework'` to `'homework`', which (of course) is always false. To make matters worse, if an empty list is compared against an empty list, it was returning true.

I simply made the user-defined labels lowercase and added a sanity check to make sure we're not comparing an empty list, and it seems to work fine now.

Jay2645 committed:
Fixed linter error.

MartinHjelmare commented on homeassistant/components/calendar/todoist.py:
```diff
@@ -365,7 +365,7 @@ def create_todoist_task(self, data):
         if not any(label in task[LABELS]
                    for label in self._label_whitelist) and (
-                   len(self._label_whitelist) > 0):
+                           len(self._label_whitelist) > 0):
```
Don't check the length explicitly, just:
```py
if ... and self._label_whitelist:
```
An empty whitelist will evaluate to `False`.

MartinHjelmare replied:
Even better actually would be to turn the expression around and check the whitelist first, then compare the whitelist labels with the task labels.

MartinHjelmare reviewed (changes requested):
Last clean up, then this should be good to go.

Jay2645 committed:
Stopped checking whitelist length explicitly.

MartinHjelmare reviewed (approved):
Nice!
